File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.ts
function maximumSum(nums: number[], m: number, l: number, r: number): number {

    const unboundDP = (pre: number[], n: number, l: number, r: number): number => {
        const dp: number[] = new Array(n + 1).fill(0);
        const dq: number[] = [];
        let ans = -1e18;

        for (let i = l; i <= n; i++) {
            const t = i - l;
            const cur = dp[t] - pre[t];

            while (
                dq.length &&
                dp[dq[dq.length - 1]] - pre[dq[dq.length - 1]] <= cur
            ) {
                dq.pop();
            }

            dq.push(t);

            while (dq.length && dq[0] < i - r) {
                dq.shift();
            }

            dp[i] = dp[i - 1];

            if (dq.length) {
                const v = pre[i] + dp[dq[0]] - pre[dq[0]];
                dp[i] = Math.max(dp[i], v);
                ans = Math.max(ans, v);
            }
        }

        return ans;
    };

    const n = nums.length;
    const pre: number[] = new Array(n + 1).fill(0);

    for (let i = 0; i < n; i++) {
        pre[i + 1] = pre[i] + nums[i];
    }

    const neg = -1e18;

    if (m >= Math.floor(n / l)) {
        return unboundDP(pre, n, l, r);
    }

    let prev: number[] = new Array(n + 1).fill(0);
    let ans = neg;
    let last = neg;

    for (let k = 1; k <= m; k++) {
        const cur: number[] = new Array(n + 1).fill(neg);
        const dq: number[] = [];

        for (let i = l; i <= n; i++) {
            const t = i - l;
            const val = prev[t] - pre[t];

            while (
                dq.length &&
                prev[dq[dq.length - 1]] - pre[dq[dq.length - 1]] <= val
            ) {
                dq.pop();
            }

            dq.push(t);

            while (dq.length && dq[0] < i - r) {
                dq.shift();
            }

            if (dq.length) {
                cur[i] = pre[i] + prev[dq[0]] - pre[dq[0]];
                ans = Math.max(ans, cur[i]);
            }
        }

        if (ans === last) {
            break;
        }

        last = ans;

        const best: number[] = new Array(n + 1).fill(neg);

        for (let i = 1; i <= n; i++) {
            best[i] = Math.max(best[i - 1], cur[i]);
        }

        prev = best;
    }

    return ans;
}

File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.rs
use std::collections::VecDeque;

impl Solution {
    pub fn maximum_sum(nums: Vec<i32>, m: i32, l: i32, r: i32) -> i64 {
        fn unbound_dp(pre: &[i64], n: usize, l: usize, r: usize) -> i64 {
            let mut dp = vec![0i64; n + 1];
            let mut dq: VecDeque<usize> = VecDeque::new();
            let mut ans = i64::MIN / 2;

            for i in l..=n {
                let t = i - l;
                let cur = dp[t] - pre[t];

                while let Some(&x) = dq.back() {
                    if dp[x] - pre[x] <= cur {
                        dq.pop_back();
                    } else {
                        break;
                    }
                }

                dq.push_back(t);

                while let Some(&x) = dq.front() {
                    if x < i.saturating_sub(r) {
                        dq.pop_front();
                    } else {
                        break;
                    }
                }

                dp[i] = dp[i - 1];

                if let Some(&x) = dq.front() {
                    let v = pre[i] + dp[x] - pre[x];
                    dp[i] = dp[i].max(v);
                    ans = ans.max(v);
                }
            }

            ans
        }

        let n = nums.len();
        let l = l as usize;
        let r = r as usize;
        let m = m as usize;

        let mut pre = vec![0i64; n + 1];
        for i in 0..n {
            pre[i + 1] = pre[i] + nums[i] as i64;
        }

        let neg = i64::MIN / 2;

        if m >= n / l {
            return unbound_dp(&pre, n, l, r);
        }

        let mut prev = vec![0i64; n + 1];
        let mut ans = neg;
        let mut last = neg;

        for _ in 0..m {
            let mut cur = vec![neg; n + 1];
            let mut dq: VecDeque<usize> = VecDeque::new();

            for i in l..=n {
                let t = i - l;
                let val = prev[t] - pre[t];

                while let Some(&x) = dq.back() {
                    if prev[x] - pre[x] <= val {
                        dq.pop_back();
                    } else {
                        break;
                    }
                }

                dq.push_back(t);

                while let Some(&x) = dq.front() {
                    if x < i.saturating_sub(r) {
                        dq.pop_front();
                    } else {
                        break;
                    }
                }

                if let Some(&x) = dq.front() {
                    cur[i] = pre[i] + prev[x] - pre[x];
                    ans = ans.max(cur[i]);
                }
            }

            if ans == last {
                break;
            }

            last = ans;

            let mut best = vec![neg; n + 1];
            for i in 1..=n {
                best[i] = best[i - 1].max(cur[i]);
            }

            prev = best;
        }

        ans
    }
}